dot/sync/chain_sync.go

```diff
@@ -691,17 +712,43 @@ func handleReadyBlock(bd *types.BlockData, pendingBlocks DisjointBlockSet, ready
 }
 
 // determineSyncPeers returns a list of peers that likely have the blocks in the given block request.
-func (cs *chainSync) determineSyncPeers(_ *network.BlockRequestMessage) []peer.ID {
+func (cs *chainSync) determineSyncPeers(req *network.BlockRequestMessage, peersTried map[peer.ID]struct{}) []peer.ID {
+	var start uint64
+	if req.StartingBlock.IsUint64() {
+		start = req.StartingBlock.Uint64()
+	}
+
 	cs.RLock()
 	defer cs.RUnlock()
 
+	// if we're currently ignoring all our peers, clear out the list.
+	if len(cs.peerState) == len(cs.ignorePeers) {
+		cs.RUnlock()
+		cs.Lock()
+		for p := range cs.ignorePeers {
+			delete(cs.ignorePeers, p)
+		}
+		cs.Unlock()
+		cs.RLock()
+	}
+
 	peers := make([]peer.ID, 0, len(cs.peerState))
 
-	for p := range cs.peerState {
+	for p, state := range cs.peerState {
 		if _, has := cs.ignorePeers[p]; has {
 			continue
 		}
 
+		if _, has := peersTried[p]; has {
+			continue
+		}
+
+		// if peer definitely doesn't have any blocks we want in the request,
+		// don't request from them
```

Comment:
should this comment be after the last if statement? Also, the comment can explain the condition to remove peer that does not have the sought number

Reply:
i think this should be here, as it's above the case that checks the highest block the peer has. basically if a peer reports some highest number x, but we want to request  x+1 for example, we shouldn't request from that peer as they won't have it. I can add more info to it